File: 14 Cluster Analysis/MNIST Cluster.py
import numpy as np
import tensorflow as tf
from sklearn.metrics import confusion_matrix
from datetime import datetime as dt
from KMeansCluster import best_km

((X_train, y_train), 
 (X_test, y_test)) = tf.keras.datasets.mnist.load_data()

N_train = len(X_train)                  # 60000 training samples
N_test = len(X_test)                    # 10000 test samples
X_train = X_train.reshape(N_train, -1).astype('float32') / 255.0
X_test = X_test.reshape(N_test, -1).astype('float32') / 255.0
K = 10

#%%
start = dt.now()
_, km = best_km(X_train, K)                     # K-means with K=10
y_hat_km = km.predict(X_test)
tab = confusion_matrix(y_hat_km, y_test)        # Confusion matrix
print(tab)

pos_dict = dict(zip(np.arange(K), np.argmax(tab, axis=1)))
print(pos_dict)
y_pred = [pos_dict[k] for k in y_hat_km]
tab = confusion_matrix(y_pred, y_test)          # Confusion matrix
print(tab)
print(sum(tab.diagonal()) / len(y_test))        # Accuracy
print(dt.now() - start)                         # K-means

#%%
from sklearn.decomposition import PCA

start = dt.now()
pca = PCA(n_components=.95)
X_train_pca = pca.fit_transform(X_train)
X_test_pca = pca.transform(X_test)
print(f'Total number of components: {pca.n_components_}')

_, km_pca = best_km(X_train_pca, K)             # K-means with K=10
y_hat_pca = km_pca.predict(X_test_pca)
tab = confusion_matrix(y_hat_pca, y_test)       # Confusion matrix
print(tab)

pos_dict = dict(zip(np.arange(K), np.argmax(tab, axis=1)))
print(pos_dict)
y_pred = [pos_dict[k] for k in y_hat_pca]
tab = confusion_matrix(y_pred, y_test)          # Confusion matrix
print(tab)
print(sum(tab.diagonal()) / len(y_test))        # Accuracy
print(dt.now() - start)                         # PCA + K-means

#%%
from sklearn.neighbors import KNeighborsClassifier

start = dt.now()
knn = KNeighborsClassifier(n_neighbors=3)       # KNN with K=3
knn.fit(X_train, y_train)
y_hat_knn = knn.predict(X_test)
tab = confusion_matrix(y_hat_knn, y_test)       # Confusion matrix
print(tab)
print(sum(tab.diagonal()) / len(y_test))        # Accuracy
print(dt.now() - start)                         # KNN

#%%
start = dt.now()
knn_pca = KNeighborsClassifier(n_neighbors=3)   # KNN with K=3
knn_pca.fit(X_train_pca, y_train)
y_hat_pca = knn_pca.predict(X_test_pca)
tab = confusion_matrix(y_hat_pca, y_test)       # Confusion matrix
print(tab)
print(sum(tab.diagonal()) / len(y_test))        # Accuracy
print(dt.now() - start)                         # PCA + KNN

File: 14 Cluster Analysis/MNIST Cluster.R
#Set directory: Run this on source instead of Console!!
setwd(dirname(rstudioapi::getActiveDocumentContext()$path))

################################################################################
library(keras)        # MNIST dataset
library(nnet)
source("KMeansCluster.R")

# Load MNIST dataset
mnist <- dataset_mnist()
X_train <- mnist$train$x / 255
y_train <- mnist$train$y
X_test <- mnist$test$x / 255
y_test <- mnist$test$y

X_train <- matrix(X_train, nrow=length(y_train))
X_test <- matrix(X_test, nrow=length(y_test))
K <- 10

predict.kmeans <- function(object, newdata) {
  # Squared Euclidean distance of each sample to each cluster center
  sd_by_center <- apply(object$centers, 1, function(x) {
    colSums((t(newdata) - x)^2)
  })
  apply(sd_by_center, 1, which.min)
}

################################################################################
y_test[y_test == 0] <- K

start <- Sys.time()
km <- best_km(X_train, K)             # K-means with K=10
y_hat_kmeans <- predict.kmeans(km, newdata=X_test)
(tab <- table(y_hat_kmeans, y_test))  # Confusion matrix
(pos_dict <- apply(tab, 1, which.max))
y_pred <- pos_dict[as.character(y_hat_kmeans)]
(tab <- table(y_pred, y_test))        # Confusion matrix
mean(y_pred == y_test)                # Accuracy
Sys.time() - start                    # K-means

################################################################################
start <- Sys.time()
pca <- prcomp(X_train, center=TRUE, scale=FALSE)
s <- pca$sdev
t <- sum(s^2)           # Compute total variance
cumvar <- cumsum(s^2/t) # Cumulative sum of proportion of variance
(idx <- which(cumvar >= 0.95)[1])

X_train_pca <- predict(pca, newdata=X_train)[,1:idx]
X_test_pca <- predict(pca, newdata=X_test)[,1:idx]
km_pca <- best_km(X_train_pca, K)     # K-means with K=10
y_hat_pca <- predict.kmeans(km_pca, newdata=X_test_pca)
(tab <- table(y_hat_pca, y_test))     # Confusion matrix
(pos_dict <- apply(tab, 1, which.max))
y_pred <- pos_dict[as.character(y_hat_pca)]
(tab <- table(y_pred, y_test))        # Confusion matrix
mean(y_pred == y_test)                # Accuracy
Sys.time() - start                    # PCA + K-means

################################################################################
library(class)
y_test[y_test == K] <- 0

start <- Sys.time()
y_hat_knn <- knn(train=X_train, test=X_test, 
                 cl=factor(y_train), k=3) # KNN with K=3
(tab <- table(y_hat_knn, y_test))     # Confusion matrix
mean(y_hat_knn == y_test)             # Accuracy
Sys.time() - start                    # KNN

################################################################################
start <- Sys.time()
y_hat_pca <- knn(train=X_train_pca, test=X_test_pca, 
                 cl=factor(y_train), k=3) # KNN with K=3
(tab <- table(y_hat_pca, y_test))     # Confusion matrix
mean(y_hat_pca == y_test)             # Accuracy
Sys.time() - start                    # PCA + KNN